src/mem/cache/replacement_policies/sim_object_unit_tester.cc

```diff
@@ -0,0 +1,193 @@
+
+#include "base/trace.hh"
+#include "mem/cache/replacement_policies/fifo_rp.hh"
+#include "mem/cache/replacement_policies/replaceable_entry.hh"
+
+namespace gem5
+{
+namespace replacement_policy
+{
+
+SimObjectUnitTester::SimObjectUnitTester(
+  const SimObjectUnitTesterParams& params):
+  SimObject(params),
+  nextEvent([this](){ processNextEvent(); }, name() + "nextEvent"),
+  replacementPolicy(params.replacement_policy),
+  numEntries(params.num_entries),
+  replacementPolicyName(params.name)
+{}
+
+void
+SimObjectUnitTester::processNextEvent()
+{
+  std::cout << "tick: " << curTick() <<
+  ", Hello from SimObjectUnitTester::processNextEvent!" << std::endl;
+
+  // create a new entry and put it in candidates
+  ReplaceableEntry* temp = new ReplaceableEntry();
+  temp->replacementData = replacementPolicy->instantiateEntry();
+  replacementPolicy->reset(temp->replacementData); // record insertion tick
+  // std::cout<< "Inserted tick: " <<
+  //std::static_pointer_cast<FIFO::FIFOReplData>(temp->replacementData)
+  //->tickInserted <<std::endl;
+  candidates.push_back(temp);
+
+  if (numEntries > 0){
+    schedule(nextEvent, curTick() + 500);
+    numEntries--;
+  }
+  else {
+    bool correct = checkCorrectness(replacementPolicy);
+    // free memory
+    freeCandidates();
+    if (correct){
+      exit(0);
+    }
+    exit(1);
+  }
+}
+
+void
+SimObjectUnitTester::freeCandidates(void){
+  for (const auto& candidate : candidates) {
+    delete candidate;
+  }
+}
+
+// bool
+// SimObjectUnitTester::checkCorrectness(void){
+//   return false;
+// }
+
+
+bool
+SimObjectUnitTester::checkCorrectness(FIFO*){
+// SimObjectUnitTester::checkCorrectnessFIFO(){
+
+  ReplaceableEntry* victim = replacementPolicy->getVictim(candidates);
+  if (std::static_pointer_cast<FIFO::FIFOReplData>(
+    victim ->replacementData)->tickInserted != 1){
+
+    std::cout<<"tick: "<<std::static_pointer_cast<FIFO::FIFOReplData>(
+      victim ->replacementData)->tickInserted <<std::endl;
+    return false;
+  }
+  return true;
+}
+
+bool
+SimObjectUnitTester::checkCorrectness(LRU*){
+
+// SimObjectUnitTester::checkCorrectnessLRU(){
+  // touch the first entry. This way, the second entry to be instantiated will
+  // be evicted.
+  replacementPolicy->touch(candidates[0]->replacementData);
+  ReplaceableEntry* victim = replacementPolicy->getVictim(candidates);
+  Tick evictedLastTick = std::static_pointer_cast<LRU::LRUReplData>(
+    victim->replacementData)->lastTouchTick;
+  Tick expectedEvictedLastTick = std::static_pointer_cast<LRU::LRUReplData>(
+    candidates[1]->replacementData)->lastTouchTick;
+  // for (const auto& candidate : candidates){
+  //   if (std::static_pointer_cast<LRU::LRUReplData>(
+  //     candidate->replacementData)->lastTouchTick > evictedLastTick){
+  //     return false;
+  //   }
+  // }
+  if (evictedLastTick != expectedEvictedLastTick){
+    return false;
+  }
+  return true;
+}
+
+
+
+bool
+SimObjectUnitTester::checkCorrectness(Base*){
```

Comment:
There's probably a better way of calling different functions based on the replacement policy type; I just don't know what that would be. I would really appreciate help/feedback on this @BobbyRBruce!

Reply:
I think the way to do this is have the `ReplacementPolicyTester` be a base class with an abstract `checkCorrectness` function. Then for each replacement policy you have a subclass which implements this function (e.g., you'd have a `FIFOReplacementPolicyTester` with its own `checkCorrectness` function that overrides that in `ReplacementPolicyTester`). Everything general to all replacement policies in the ReplacementPolicyTester and anything specific to each implementation in their respective subclasses.